### DISCUSSION TITLE
Rethinking the 'Apply' & 'Clear' functionality in Combobox (Filters)

### DISCUSSION THREAD

**Discussion by timdhahn** [microsoft-search/pnp-modern-search#2011] [Ideas]
**Is your feature request related to a problem? Please describe.**
Accessibility is currently an issue with ComboBox. It is not possible to reach the "Apply" and "Clear" button when filtering via keyboard only. You can generally reach each checkbox, but can not reach the sticky "Apply" and "Clear" buttons. The same issue happens with the 'AND/OR' selector.

An unintended consequence of having the buttons in the dropdown means they do not show when filtering. 

A better experience might be changing how the Filters are applied.

![image](https://user-images.githubusercontent.com/7631802/162273912-fbcaf642-8598-4439-aed8-f5beaa0eec70.png)

**Describe the solution you'd like**
Moving the "Apply" and "Clear" buttons outside of each filter and consolidate them below the filters.

![image](https://user-images.githubusercontent.com/7631802/162271581-7f0312a8-3954-43e5-8db5-da4f3da2b69b.png)

**Describe alternatives you've considered**
Another alternative would be to make the "Apply" and "Clear" accessible (along with the And/Or).  After digging into some code it seems like ComboBox isn't made to support accessible buttons inside its dropdown. The chooses given are 'Normal', 'Header', 'Divider'. With these choices it's not possible to reach buttons or other functionality besides checkboxes/options inside the the dropdown via the keyboard.

Alternatively, filters could automatically be applied after selection or when the dropdown is closed.

Anyone have other ideas on how to best approach this issue?

**Comment by FranckyC** [comment]
Hi @timdhahn, thank you for the feedback. Unfortunately, the behavior you describe is not so simple to implement due to the internal architecture used and native SharePoint filter behavior. The filters can't be grouped together and use a single Apply/Clear button. The Handlebars logic would be too complicated.

I'll check for the keyboard navigation but at first glance, not simple also as we use Office UI Fabric components.

BTW I see you don't use a recent version as the buttons appear at the bottom of the list. In latest versions, buttons appeart at the top.

**Reply by Neko-Ferro** [reply]
Excluding all users who rely on assistive technology because "it is not so simple" is not an ideal solution, or a solution at all.  Accessibility should be on the forefront of every interactive element developed for users. Ableism should not be a foundation of everything that is coded for the community. It should be relatively simple to ensure that anything that is created have a visible focus and is reachable via keyboard.

The fact that something was coded in such a way that it is now an unmaking soup scenario to ensure simple, basic accessibility shows a troubling lack of awareness of the needs of those who rely on assistive technology, and opens up companies who adopt these solutions to serious liability.

**Reply by FranckyC** [reply]
@Neko-Ferro just a quick reminder this is an open source project made on my spare time. I'm not a company and PnP Modern Search is not a Microsoft product. I didn't say that was impossible but that it will take some time to figure out how to do it since we are using Office Fluent components so we don't own all the DOM elements. 

> The fact that something was coded in such a way that it is now an unmaking soup scenario to ensure simple, basic accessibility shows a troubling lack of awareness of the needs of those who rely on assistive technology, and opens up companies who adopt these solutions to serious liability.

Did you review the code? Did you try to update it to correct this behavior and make it better? Instead of blaming, you should probably start here.

**Reply by wobba** [reply]
@Neko-Ferro it's harder than you think when it comes to the UI framework components. And we don't disagree that accessibility is not important. That said, those who push for accessibility can also contribute to make it better for everyone. For many of the components in the web parts you can provide your own handlebar templates with the markup you need as well. And if you do, please contribute back :)
